### PR TITLE
Fix pool preallocation tests

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
It uses a more reliable way to determine whether pool preallocation was successful.
Previously, the test measured free device memory, but it depends too heavily on driver behavior.
The test has been reworked so that it uses the default resource's pool interface directly and tries to allocate an increasing amount memory from the pool (without allocating more from the OS). When the request fails, a preallocation request is issued and the allocation is retried. At that point it's expected to succeed.

### Tests:
(Old tests removed / modified)
- [ ] Existing tests apply
- [X] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Review by jantonguirao — APPROVED** [review]

File: dali/core/mm/default_resource_test.cc [hunk 1]
```diff
@@ -374,54 +388,88 @@ TEST(MMDefaultResource, DISABLED_ReleaseUnusedMaxMem) {
...
+  void *mem = pool->try_allocate_from_free(size, alignment);
+  EXPECT_NE(mem, nullptr) << "Preallocation succeeded, so we should be able to get the "
+                             "requested amount of memory from the pool.";
+
+  res->deallocate(mem, size, alignment);
+
+  mm::ReleaseUnusedMemory();
+}
+
+TEST(MMDefaultResource, PreallocateDeviceMemory) {
```
Comment (JanuszL) [inline dali/core/mm/default_resource_test.cc:427]:
```suggestion
TEST(MMDefaultResource, MultiGPUPreallocateDeviceMemory) {
```
So we run it as a part of `TL0_multigpu`

File: dali/core/mm/default_resource_test.cc [hunk 2]
```diff
@@ -374,54 +388,88 @@ TEST(MMDefaultResource, DISABLED_ReleaseUnusedMaxMem) {
   ReleaseUnusedTestImpl();
 }
 
-static void PreallocateTestImpl() {
+TEST(MMDefaultResource, PreallocatePinnedMemory) {
+  mm::ReleaseUnusedMemory();  // release any unused memory to check that we're really preallocating
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/core/mm/default_resource_test.cc:392]: Well... assuming that no memory in the pool is actually in use.

File: dali/core/mm/default_resource_test.cc (hunk 1 above)
Comment (mzient) [reply dali/core/mm/default_resource_test.cc:427]: It's not _that_ important. I wanted some extra coverage (and indeed it helped uncover some old bugs), but the test has its merit also when there's only one GPU.

File: dali/core/mm/default_resource_test.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/core/mm/default_resource_test.cc:392]: I can't imaging how any memory could be in use as this is the begging of the test.

**Review by JanuszL — APPROVED** [review]

**Pull request merged**